Implement ReflectionContext::addImage with section finder

augusto2112 opened the pull request:
Implement a version of addImage that takes in a closure responsible
for finding the sections of interest. This allows for the registration
of metadata even in situations where the object file is not complete
(for example, when generated by the JIT).

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on lib/Basic/ObjectFile.cpp:
```diff
@@ -0,0 +1,21 @@
+//===--- ObjectFile.cpp - Object File Related Information ------*- C++ -*-===//
```
FYI, the `-*- C++ -*-` marker only makes sense in a `.h` file, where it's ambiguous whether it's a C or a C++ header.

adrian-prantl commented on lib/Basic/ObjectFile.cpp:
```diff
@@ -0,0 +1,21 @@
+
+using namespace swift;
+
+SwiftObjectFileFormat &SwiftObjectFileFormat::getSwiftObjectFileFormat() {
```
Are we sure that a static implementation is good enough for all use-cases? Is this function still useful if we are, for example on a Mac and debug a Linux process over the network?

augusto2112 replied:
Ok! I changed this so the file format is resolved on LLDB's side, by consulting the module's triple's format type.

adrian-prantl replied:
Does that mean we can remove this function?

kastiglione commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,46 @@ class ReflectionContext
+        ReflectionSectionKind::builtin, ReflectionSectionKind::capture,
+        ReflectionSectionKind::typeref, ReflectionSectionKind::reflstr};
+
+    std::vector<std::pair<RemoteRef<void>, uint64_t>> Pairs;
```
with a known fixed size, this could be `llvm::SmallVector<T, 6>`

augusto2112 replied:
Nice, thanks!

kastiglione commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,46 @@ class ReflectionContext
+  ///     /b True if any of the reflection sections were registered,
+  ///     /b false otherwise.
+  bool
+  addImage(std::function<std::pair<RemoteRef<void>, uint64_t>(llvm::StringRef)>
```
this could be a `llvm::function_ref` to indicate it's a non-escaping function/lambda.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,46 @@ class ReflectionContext
+      }
+    }
+
+    if (llvm::all_of(Pairs, [](const auto &Pair) { return !Pair.first; }))
```
I think this could benefit from a comment: Are we requiring that all sections are present here?

adrian-prantl replied:
Also, perhaps this could just be an early-exit in the loop above?

augusto2112 replied:
It requires that at least one section be present, so it can't be an early exit

adrian-prantl replied:
Makes sense now. I stopped reading after `all_of`.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,46 @@ class ReflectionContext
+    std::vector<std::pair<RemoteRef<void>, uint64_t>> Pairs;
+    for (auto Section : Sections) {
+      Pairs.push_back(FindSection(ObjectFileFormat.getSectionName(Section)));
+      if (Pairs.back().first) {
```
since we use `Pairs.back().first` multiple times and it's not obvious what it is, perhaps write this as `auto descriptive_name = Pairs.back().first;`?

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,47 @@ class ReflectionContext
+  ///     of freeing the memory buffer in the RemoteRef return value.
+  ///     process.
+  /// \return
+  ///     /b True if any of the reflection sections were registered,
```
I'm not super familiar with Doxygen conventions so maybe this is correct: is this supposed to be a forward-slash in `/b`?

augusto2112 replied:
Tested it out and you're right, it should be `/b`

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -644,6 +645,47 @@ class ReflectionContext
+      return false;
+
+    ReflectionInfo Info = {
+        {Pairs[0].first, Pairs[0].second}, {Pairs[1].first, Pairs[1].second},
```
Are those pairs, too. Ie.: Would `ReflectionInfo Info = {Pairs[0], Pairs[1], ...}` work, too?

augusto2112 replied:
That doesn't work, unfortunately :(

kastiglione replied:
I looked at this too, they're not pairs, unfortunately. It's too bad we don't have a way to construct from a pair, like `{*pair}` expanding to `{pair.first, pair.second}`.

augusto2112 committed:
Implement ReflectionContext::addImage with section finder

Implement a version of addImage that takes in a closure responsible
for finding the sections of interest. This allows for the registration
of metadata even in situations where the object file is not complete
(for example, when generated by the JIT).

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
Failures are unrelated:
```
16:05:45 Failed Tests (2):
16:05:45   Swift(macosx-x86_64) :: Interpreter/SDK/objc_bridge_cast.swift
16:05:45   Swift(macosx-x86_64) :: Interpreter/SDK/objc_dealloc.swift
```

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
Unrelated failure:

```
11:26:08 
<unknown>:0: error: module file '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-main/buildbot_linux/xctest-linux-x86_64/module-cache/199U4XZW60V9Z/SwiftShims-FOQQR403J715.pcm' is out of date and needs to be rebuilt: signature mismatch
```

augusto2112 commented:
@swift-ci smoke test
